internal/ihop/creator.go

```diff
@@ -0,0 +1,186 @@
+package ihop
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/paketo-buildpacks/packit/v2/scribe"
+)
+
+//go:generate faux --interface ImageClient --output fakes/image_client.go
+type ImageClient interface {
+	Build(definitionImage DefinitionImage, platform string) (Image, error)
+	Update(Image) (Image, error)
+}
+
+//go:generate faux --interface ImageBuilder --output fakes/image_builder.go
+type ImageBuilder interface {
+	Execute(definitionImage DefinitionImage, platform string) ImageBuildPromise
+}
+
+//go:generate faux --interface LayerCreator --output fakes/layer_creator.go
+type LayerCreator interface {
+	Create(Image, DefinitionImage, SBOM) (Layer, error)
+}
+
+type Stack struct {
+	Build []Image
+	Run   []Image
+}
+
+type Creator struct {
+	docker           ImageClient
+	builder          ImageBuilder
+	userLayerCreator LayerCreator
+	sbomLayerCreator LayerCreator
+	now              func() time.Time
+	logger           scribe.Logger
+}
+
+func NewCreator(docker ImageClient, builder ImageBuilder, userLayerCreator, sbomLayerCreator LayerCreator, now func() time.Time, logger scribe.Logger) Creator {
+	return Creator{
+		docker:           docker,
+		builder:          builder,
+		userLayerCreator: userLayerCreator,
+		sbomLayerCreator: sbomLayerCreator,
+		now:              now,
+		logger:           logger,
+	}
+}
+
+func (c Creator) Execute(def Definition) (Stack, error) {
+	c.logger.Title("Building %s", def.ID)
+
+	var stack Stack
+	for _, platform := range def.Platforms {
+		c.logger.Process("Building on %s", platform)
+
+		build, run, err := c.create(def, platform)
+		if err != nil {
+			return Stack{}, err
+		}
+
+		stack.Build = append(stack.Build, build)
+		stack.Run = append(stack.Run, run)
+	}
+
+	return stack, nil
+}
+
+func (c Creator) create(def Definition, platform string) (Image, Image, error) {
+	c.logger.Subprocess("Building base images")
+
+	buildPromise := c.builder.Execute(def.Build, platform)
+	runPromise := c.builder.Execute(def.Run, platform)
+
+	build, buildSBOM, err := buildPromise.Resolve()
+	if err != nil {
+		return Image{}, Image{}, err
+	}
+
+	run, runSBOM, err := runPromise.Resolve()
+	if err != nil {
+		return Image{}, Image{}, err
+	}
+
+	c.logger.Action("Build complete for base images")
+
+	packages := NewPackages(buildSBOM.Packages(), runSBOM.Packages())
+	timestamp := c.now()
+
+	c.logger.Subprocess("build: Decorating base image")
+
+	c.logger.Action("Adding CNB_* environment variables")
+	build.Env = append(build.Env, fmt.Sprintf("CNB_USER_ID=%d", def.Build.UID))
+	build.Env = append(build.Env, fmt.Sprintf("CNB_GROUP_ID=%d", def.Build.GID))
+	build.Env = append(build.Env, fmt.Sprintf("CNB_STACK_ID=%s", def.ID))
```

Comment:
This might be a silly question, but are these environment variables only set on the build image?

Reply:
The stack id is also used in the labels that get added to both images: https://github.com/paketo-buildpacks/jam/pull/53/files#diff-fd20e32de0ce73c2aa1b986218a8fc41a292cf3db3f0e95b054e987a49b911beR141